diff --git a/panda/board/drivers/can_common.h b/panda/board/drivers/can_common.h
index 35c6702e..73d51517 100644
--- a/panda/board/drivers/can_common.h
+++ b/panda/board/drivers/can_common.h
@@ -158,6 +158,8 @@ void can_set_forwarding(uint8_t from, uint8_t to) {
 void ignition_can_hook(CANPacket_t *msg) {
   if (msg->bus == 0U) {
     int len = GET_LEN(msg);
+    const int TESLA_DI_GEAR_P = 1;
+    static int tesla_gear = -1;
 
     // GM exception
     if ((msg->addr == 0x1F1U) && (len == 8)) {
@@ -186,15 +188,25 @@ void ignition_can_hook(CANPacket_t *msg) {
       int counter = msg->data[6] >> 4;
 
       static int prev_counter_tesla = -1;
+      static int prev_power_state_tesla = 0x7F;
       if ((counter == ((prev_counter_tesla + 1) % 16)) && (prev_counter_tesla != -1)) {
         // VCFRONT_LVPowerState->VCFRONT_vehiclePowerState
         int power_state = (msg->data[0] >> 5U) & 0x3U;
-        ignition_can = power_state == 0x3;  // VEHICLE_POWER_STATE_DRIVE=3
+        bool power_state_increased = power_state > prev_power_state_tesla;
+        bool power_state_drive = power_state == 0x3;  // VEHICLE_POWER_STATE_DRIVE=3
+        bool tesla_not_in_park = (tesla_gear != TESLA_DI_GEAR_P) && (tesla_gear != -1);
+        ignition_can = power_state_increased || (power_state_drive && tesla_not_in_park);
         ignition_can_cnt = 0U;
+        prev_power_state_tesla = power_state;
       }
       prev_counter_tesla = counter;
     }
 
+    if ((msg->addr == 0x118U) && (len == 8)) {
+      // broadcasted only in VEHICLE_POWER_STATE_DRIVE
+      tesla_gear = (msg->data[2] >> 5) & 0x7;
+    }
+
     // Mazda exception
     if ((msg->addr == 0x9EU) && (len == 8)) {
       ignition_can = (msg->data[0] >> 5) == 0x6U;
